overcooked_ai_py/mdp/overcooked_mdp.py

```diff
@@ -669,27 +696,151 @@ def get_state_transition(self, state, joint_action):
+
+    def get_full_pots(self, pot_states):
+        return self.get_cooking_pots(pot_states) + self.get_ready_pots(pot_states)
+
+    def get_partially_full_pots(self, pot_states):
+        return pot_states["tomato"]["partially_full"] + pot_states["onion"]["partially_full"]
+
+    def is_dish_pickup_useful(self, state, pot_states):
+        """
+        NOTE: this only works if self.num_players == 2
+        Useful if:
+        - Pot is ready/cooking and there is no player with a dish               \ 
+        - 2 pots are ready/cooking and there is one player with a dish          | -> number of dishes in players hands < number of ready/cooking/partially full soups 
+        - Partially full pot is ok if the other player is on course to fill it  /
+
+        We also want to prevent picking up and dropping dishes, so add the condition
+        that there must be no dishes on counters
+        """
+        # This next line is to prevent reward hacking (this logic is also used by reward shaping)
+        dishes_on_counters = self.get_counter_objects_dict(state)["dish"]
+        no_dishes_on_counters = len(dishes_on_counters) == 0
+
+        num_player_dishes = len(state.player_objects_by_type['dish'])
+        non_empty_pots = len(self.get_ready_pots(pot_states) + self.get_cooking_pots(pot_states) + self.get_partially_full_pots(pot_states))
+        return no_dishes_on_counters and num_player_dishes < non_empty_pots
+
+    def is_dish_drop_useful(self, state, pot_states, player_index):
+        """
+        NOTE: this only works if self.num_players == 2
+        Useful if:
+        - Onion is needed (all pots are non-full)
+        - Nobody is holding onions
+        """
+        if self.num_players != 2: return False
+        all_non_full = len(self.get_full_pots(pot_states)) == 0
+        other_player = state.players[1 - player_index]
+        other_player_holding_onion = other_player.has_object() and other_player.get_object().name == "onion"
+        return all_non_full and not other_player_holding_onion
+
+    def is_onion_pickup_useful(self, state, pot_states, player_index):
+        """
+        NOTE: this only works if self.num_players == 2
+        Always useful unless:
+        - All pots are full & other agent is not holding a dish
+        """
+        if self.num_players != 2: return False
+        all_pots_full = self.num_pots == len(self.get_full_pots(pot_states))
+        other_player = state.players[1 - player_index]
+        other_player_has_dish = other_player.has_object() and other_player.get_object().name == "dish"
+        if all_pots_full and not other_player_has_dish:
+            return False
+        return True
+        
+    def is_onion_drop_useful(self, state, pot_states, player_index):
+        """
+        NOTE: this only works if self.num_players == 2
+        Useful if:
+        - Dish is needed (all pots are full)
+        - Nobody is holding a dish
```

Comment:
There could be some rare cases that don't fit into the criteria for these various events being useful. E.g. An onion drop can be useful if the other player has an onion and is closer. But I don't think these rare cases would change the overall stats much so they probably don't matter.

Reply:
Yeah, I think the "useful" stats should all be taken as approximate/with a grain of salt. I mainly want to use it as a diagnostic of "if i see that <20% of X drops are useful", that's probably a bad sign / a sign that i should check what is going on

Reply:
Yeh that sounds right